### PR TITLE
Add Visual Information Fidelity 

### PR DESCRIPTION
**Pull request by bojobo**
## What does this PR do?

Adds metric visual information fidelity (vif). See #799 

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
:books: Documentation preview :books:: https://torchmetrics--1830.org.readthedocs.build/en/1830/

### CODE REVIEW

**Comment by SkafteNicki** [comment]
@bojobo cool adding, feel free to reach out if you need help with anything :]

**Comment by bojobo** [comment]
Will do that, thanks :)
I'll try to be done by next week. Is that early enough to get this metric into the next release? @SkafteNicki

**Comment by bojobo** [comment]
@SkafteNicki 

I'm sure that I'm on the right path, but I can't quite get the same results as the packages I'm using for reference.
My implementation is based on [the implementation found in `sewar`](https://github.com/andrewekhalel/sewar/blob/ac76e7bc75732fde40bb0d3908f4b6863400cc27/sewar/full_ref.py#L357) as well as on [the implementation found in `piq`](https://github.com/photosynthesis-team/piq/blob/c26ce7c78e2a77c1975cb50362befb60c356c6d5/piq/vif.py#L17). Furthermore, to make sure that the implementation is same as the paper, I've used [the published resources](https://live.ece.utexas.edu/research/quality/VIF.htm) (where [this .zip](https://live.ece.utexas.edu/research/quality/subjective.htm) contains the MatLab implementation). I've managed to create the same filter as `sewer`, but I can't get the same result.

Do you see any differences in the implementation? Or any fatal errors?

Also, I'm not quite sure how to implement the tests, since I've almost never done that (I know, I know, shame on me). I'm grateful for any form of help :)

**Comment by SkafteNicki** [comment]
Hi @bojobo, sorry for not responding sooner. 
I tried out the code in PR and cannot see a problem with the implementation. Here is the testing script I used:
```python
import sewar
import torch
import numpy as np

preds = np.random.randint(0, 255, (1, 41, 41, 1), dtype=np.uint8)
target = np.random.randint(0, 255, (1, 41, 41, 1), dtype=np.uint8)

compare = sewar.full_ref.vifp(target.squeeze(), preds.squeeze())


from torchmetrics.functional.image.vif import visual_information_fidelity

tm_res = visual_information_fidelity(
    torch.tensor(preds).permute(0,3,1,2),
    torch.tensor(target).permute(0,3,1,2),
    data_range=255.0
).item()

print(compare)
print(tm_res)
print(abs(compare - tm_res))
```
I get a difference in result of 1e-7 to 1e-9 which. So the implementation seems to be working to me.

**Comment by bojobo** [comment]
@SkafteNicki Thanks for the reply!
I have some follow-up questions.

- I don't understand why `mypy` is failing. I understand the returned message, but if I'm correct, other metrics are computed similarly and don't fail.
- `ddp=True` leads to failing tests. I don't know what I'm missing to make it work. I would appreciate any help.

**Comment by SkafteNicki** [comment]
> @SkafteNicki Thanks for the reply! I have some follow-up questions.
> 
>     * I don't understand why `mypy` is failing. I understand the returned message, but if I'm correct, other metrics are computed similarly and don't fail.
> 
>     * `ddp=True` leads to failing tests. I don't know what I'm missing to make it work. I would appreciate any help.

Hi @bojobo, I have fixed the typing issue. These two lines were missing:
https://github.com/Lightning-AI/torchmetrics/blob/45cc59192dc2401e3d591e79a447dcda51ddf188/src/torchmetrics/image/vif.py#L44-L45
which informs mypy about the type of the metric states.

For the ddp, I have to look further into that.

**Comment by SkafteNicki** [comment]
@bojobo the ddp test should also pass now.
The only problem was the default tolerance, which i have inscreased from 1e-8 to 1e-7 which is a completely fine tolerance to still be close to the reference:
https://github.com/Lightning-AI/torchmetrics/blob/b1098012115a1b37475b151fd0a8dd3361a2dd10/tests/unittests/image/test_vif.py#L52-L56

File: src/torchmetrics/image/vif.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    - ``target`` (:class:`~torch.Tensor`): Ground truth values of shape ``(N,C,H,W)``
+
+    As output of `forward` and `compute` the metric returns the following output
+
+    - ``vif-p`` (:class:`~torch.Tensor`): Tensor with vif-p score
+
+    Args:
+        sigma_n_sq: variance of the visual noise
+        kwargs: Additional keyword arguments, see :ref:`Metric kwargs` for more info.
+    """
```
Comment (SkafteNicki) [inline src/torchmetrics/image/vif.py:39]: Still need an example here, like:
https://github.com/Lightning-AI/torchmetrics/blob/ef1af350befa32fe0ad4bf9e91ec544ee0dc5e66/src/torchmetrics/image/uqi.py#L57-L64

File: src/torchmetrics/image/vif.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    vif_score: Tensor
+    total: Tensor
+
+    def __init__(self, sigma_n_sq: float = 2.0, **kwargs: Any) -> None:
+        super().__init__(**kwargs)
+
+        self.add_state("vif_score", default=tensor(0.0), dist_reduce_fx="sum")
+        self.add_state("total", default=tensor(0.0), dist_reduce_fx="sum")
+        self.sigma_n_sq = sigma_n_sq
```
Comment (SkafteNicki) [inline src/torchmetrics/image/vif.py:53]: lets add some input validation similar to:
https://github.com/Lightning-AI/torchmetrics/blob/ef1af350befa32fe0ad4bf9e91ec544ee0dc5e66/src/torchmetrics/image/rase.py#L77-L78

File: src/torchmetrics/image/vif.py (hunk 1 above)
Comment (bojobo) [reply src/torchmetrics/image/vif.py:47]: Added

File: src/torchmetrics/image/vif.py (hunk 2 above)
Comment (bojobo) [reply src/torchmetrics/image/vif.py:66]: Added

File: src/torchmetrics/image/vif.py (hunk 1 above)
Comment (bojobo) [reply src/torchmetrics/image/vif.py:46]: Okay, nevermind. The unit tests fail on building the doc because of the example result. How do I add that without breaking the test?

**Review by SkafteNicki — APPROVED** [review]

**Review by Borda — APPROVED** [review]
